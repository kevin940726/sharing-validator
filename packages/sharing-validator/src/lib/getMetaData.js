import fetch from "node-fetch";
import { Parser } from "htmlparser2";
import USER_AGENTS from "./userAgents";

async function getMetadata(
  url,
  { userAgent = USER_AGENTS.general, headers = {} } = {}
) {
  const response = await fetch(url, {
    headers: {
      "User-Agent": userAgent,
      ...headers
    }
  });

  let isTitle = false;
  let title = "";

  return new Promise(resolve => {
    const meta = [];

    const htmlParser = new Parser(
      {
        onopentag(name, attrs) {
          if (name === "meta" && (attrs.property || attrs.name)) {
            meta.push([attrs.property || attrs.name, attrs.content]);
          } else if (name === "title") {
            isTitle = true;
            title = "";
          }
        },
        onText(text) {
          if (isTitle) {
            title += text;
          }
        },
        onclosetag(name) {
          if (name === "head") {
            htmlParser.end();
          } else if (name === "title" && isTitle) {
            if (title) {
              meta.push(["title", title]);
            }
            isTitle = false;
          }
        },
        onend() {
          resolve(meta);
        }
      },
      { decodeEntities: false }
    );

    response.body.pipe(htmlParser);
  });
}

export default getMetadata;
