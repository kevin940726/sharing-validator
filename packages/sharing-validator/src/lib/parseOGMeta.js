function parseMetaData(meta) {
  const data = {};

  for (let [key, value] of meta) {
    if (!key || key === "undefined") {
      // Skip meta which doesn't have a `property` or `name` attribute
      continue;
    }

    const properties = key.split(":");

    let target = data;
    for (let property of properties.slice(0, -1)) {
      if (typeof target[property] === "string") {
        target[property] = {
          url: target[property]
        };
      } else {
        target[property] = target[property] || {};
      }

      target = target[property];

      if (Array.isArray(target)) {
        target = target.slice(-1)[0];
      }
    }

    const lastKey = properties.slice(-1)[0];
    const originalValue = Array.isArray(target[lastKey])
      ? target[lastKey][0]
      : target[lastKey];

    if (originalValue) {
      if (typeof originalValue === "string") {
        target[lastKey] = [].concat(target[lastKey], value);
      } else {
        target[lastKey] = [].concat(target[lastKey], { url: value });
      }
    } else {
      target[lastKey] = value;
    }
  }

  return data;
}

export default parseMetaData;
