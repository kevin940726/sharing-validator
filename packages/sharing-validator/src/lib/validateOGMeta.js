import validateMeta from "./validateMeta";
import {
  STRING_PATTERN,
  STRING_PATTERN_REQUIRED,
  URL_PATTERN,
  SECURE_URL_PATTERN,
  INTEGER_PATTERN,
  LOCALE_PATTERN
} from "../utils/validationPatterns";

const VALIDATE_PATTERNS = {
  // https://ogp.me/#metadata
  og: {
    // required
    url: STRING_PATTERN_REQUIRED,
    title: STRING_PATTERN_REQUIRED,
    // https://ogp.me/#types
    type: /^(music\.(song|album|playlist|radio_station)|video\.(movie|episode|tv_show|other)|article|book|profile|website)$/,

    // optional types
    locale: {
      pattern: LOCALE_PATTERN,
      alternate: {
        pattern: LOCALE_PATTERN
      }
    },
    description: STRING_PATTERN,
    site_name: STRING_PATTERN,
    determiner: content => ({
      valid: !content || /^(a|an|the|auto|)$/.test(content),
      message: `Expected determiner to be one of (a, an, the, "", auto), received ${content}`
    }),
    image: {
      pattern: URL_PATTERN,
      url: URL_PATTERN,
      secure_url: SECURE_URL_PATTERN,
      width: INTEGER_PATTERN,
      height: INTEGER_PATTERN,
      alt: (content, meta, key) => {
        const imageKey = key.slice(0, -1);

        let image = meta;

        try {
          for (let k of imageKey) {
            if (/[(\d+)]/.test(k)) {
              image = image[parseInt(k.slice(1, -1), 10)];
            } else {
              image = image[k];
            }
          }
        } catch (err) {
          // Not specified
          return {
            valid: true
          };
        }

        return {
          valid: !!(image.url && content),
          message:
            '"og:image.alt" is required when specifying "og:image" or "og:image:url"'
        };
      }
    },
    video: {
      pattern: URL_PATTERN,
      url: URL_PATTERN,
      secure_url: SECURE_URL_PATTERN,
      width: INTEGER_PATTERN,
      height: INTEGER_PATTERN
    },
    audio: {
      pattern: URL_PATTERN,
      url: URL_PATTERN,
      secure_url: SECURE_URL_PATTERN
    }
  }
};

function validateOGMeta(meta) {
  return validateMeta(meta, VALIDATE_PATTERNS);
}

export default validateOGMeta;
