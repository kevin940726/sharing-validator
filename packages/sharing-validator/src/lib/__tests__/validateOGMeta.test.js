import validateOGMeta from "../validateOGMeta";
import parseOGMeta from "../parseOGMeta";
import getMetaData from "../getMetaData";

test("baseline", () => {
  const meta = [
    ["undefined", "ie=edge"],
    ["viewport", "width=device-width, initial-scale=1, shrink-to-fit=no"],
    ["generator", "Gatsby 2.15.11"],
    ["description", "I&#x27;m Kai Hao, a front-end developer in Taiwan."],
    ["og:url", "https://kaihao.dev/"],
    ["og:title", "Kai Hao"],
    ["og:description", "I&#x27;m Kai Hao, a front-end developer in Taiwan."],
    [
      "og:image",
      "https://kaihao.dev/static/15baf5b2b73048942a827496d3805470/70e2c/default-meta-image.png"
    ],
    ["og:image:width", "1200"],
    ["og:image:height", "626"],
    ["og:type", "website"],
    ["twitter:card", "summary_large_image"],
    ["twitter:creator", "@kevin940726"],
    ["twitter:title", "Kai Hao"],
    [
      "twitter:description",
      "I&#x27;m Kai Hao, a front-end developer in Taiwan."
    ],
    [
      "twitter:image",
      "https://kaihao.dev/static/15baf5b2b73048942a827496d3805470/70e2c/default-meta-image.png"
    ],
    ["theme-color", "#f2994a"]
  ];

  const parsedMeta = parseOGMeta(meta);

  expect(validateOGMeta(parsedMeta)).toMatchSnapshot();
});

test("array", () => {
  const meta = [
    ["og:image", "http://example.com/rock.jpg"],
    ["og:image:width", "300"],
    ["og:image:height", "300"],
    ["og:image", "http://example.com/rock2.jpg"],
    ["og:image", "http://example.com/rock3.jpg"],
    ["og:image:height", "1000"]
  ];

  const parsedMeta = parseOGMeta(meta);

  expect(validateOGMeta(parsedMeta)).toMatchSnapshot();
});

test("spotify", async () => {
  const meta = await getMetaData(
    "https://open.spotify.com/track/4u7EnebtmKWzUH433cf5Qv"
  );

  const parsedMeta = parseOGMeta(meta);

  expect(validateOGMeta(parsedMeta)).toMatchSnapshot();
});
