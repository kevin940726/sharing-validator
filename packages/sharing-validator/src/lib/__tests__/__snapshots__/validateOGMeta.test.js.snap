// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array 1`] = `
Array [
  Object {
    "key": "og:type",
    "valid": false,
    "value": undefined,
  },
  Object {
    "key": "og:image:[0]:url",
    "valid": true,
    "value": "http://example.com/rock.jpg",
  },
  Object {
    "key": "og:image:[0]:width",
    "valid": true,
    "value": "300",
  },
  Object {
    "key": "og:image:[0]:height",
    "valid": true,
    "value": "300",
  },
  Object {
    "key": "og:image:[1]:url",
    "valid": true,
    "value": "http://example.com/rock2.jpg",
  },
  Object {
    "key": "og:image:[2]:url",
    "valid": true,
    "value": "http://example.com/rock3.jpg",
  },
  Object {
    "key": "og:image:[2]:height",
    "valid": true,
    "value": "1000",
  },
]
`;

exports[`baseline 1`] = `
Array [
  Object {
    "key": "og:url",
    "valid": true,
    "value": "https://kaihao.dev/",
  },
  Object {
    "key": "og:title",
    "valid": true,
    "value": "Kai Hao",
  },
  Object {
    "key": "og:type",
    "valid": true,
    "value": "website",
  },
  Object {
    "key": "og:description",
    "valid": true,
    "value": "I&#x27;m Kai Hao, a front-end developer in Taiwan.",
  },
  Object {
    "key": "og:image:url",
    "valid": true,
    "value": "https://kaihao.dev/static/15baf5b2b73048942a827496d3805470/70e2c/default-meta-image.png",
  },
  Object {
    "key": "og:image:width",
    "valid": true,
    "value": "1200",
  },
  Object {
    "key": "og:image:height",
    "valid": true,
    "value": "626",
  },
]
`;
